src/hist/plot.py

```diff
@@ -210,19 +218,233 @@ def plot2d_full(
+
+
+def plot_ratio(
+    _hist: hist.BaseHist,
+    ratio: np.ndarray,
+    ratio_uncert: np.ndarray,
+    ax: matplotlib.axes.Axes,
+    **kwargs: Any,
+) -> matplotlib.axes.Axes:
+    """
+    Plot a ratio plot on the given axes
+    """
+    x_values = _hist.axes[0].centers
+    left_edge = _hist.axes.edges[0][0]
+    right_edge = _hist.axes.edges[-1][-1]
+
+    # Set 0 and inf to nan to hide during plotting
+    ratio[ratio == 0] = np.nan
+    ratio[np.isinf(ratio)] = np.nan
+
+    central_value = kwargs.pop("central_value", 1.0)
+    ax.axhline(
+        central_value, color="black", linestyle="dashed", linewidth=1.0, zorder=-1
+    )
+
+    # {"line", "bar"}
+    uncert_draw_type = kwargs.pop("uncert_draw_type", "line")
+    if uncert_draw_type == "line":
+        ax.errorbar(
+            x_values,
+            ratio,
+            yerr=ratio_uncert,
+            color="black",
+            marker="o",
+            linestyle="none",
+        )
+    elif uncert_draw_type == "bar":
+        bar_width = (right_edge - left_edge) / len(ratio)
+
+        bar_top = ratio + ratio_uncert[1]
+        bar_bottom = ratio - ratio_uncert[0]
+        # bottom can't be nan
+        bar_bottom[np.isnan(bar_bottom)] = 0
+        bar_height = bar_top - bar_bottom
+
+        ax.scatter(x_values, ratio, color="black")
+        ax.bar(
+            x_values,
+            height=bar_height,
+            width=bar_width,
+            bottom=bar_bottom,
+            fill=False,
+            linewidth=0,
+            edgecolor="gray",
+            hatch=3 * "/",
+        )
+
+    ratio_ylim = kwargs.pop("ylim", None)
+    if ratio_ylim is None:
+        # plot centered around central value with a scaled view range
+        # the value _with_ the uncertainty in view is important so base
+        # view range on extrema of value +/- uncertainty
+        valid_ratios_idx = np.where(np.isnan(ratio) == False)  # noqa: E712
+        valid_ratios = ratio[valid_ratios_idx]
+        extrema = np.array(
+            [
+                valid_ratios - ratio_uncert[0][valid_ratios_idx],
+                valid_ratios + ratio_uncert[1][valid_ratios_idx],
+            ]
+        )
+        max_delta = np.max(np.abs(extrema - central_value))
+        ratio_extrema = np.abs(max_delta + central_value)
+
+        _alpha = 2.0
+        scaled_offset = max_delta + (max_delta / (_alpha * ratio_extrema))
+        ratio_ylim = [central_value - scaled_offset, central_value + scaled_offset]
+
+    ax.set_xlim(left_edge, right_edge)
+    ax.set_ylim(bottom=ratio_ylim[0], top=ratio_ylim[1])
+
+    ax.set_xlabel(_hist.axes[0].label)
+    ax.set_ylabel(kwargs.pop("ylabel", "Ratio"))
+
+    return ax
+
+
+def _plot_ratiolike(
     self: hist.BaseHist,
-    func: Union[Callable[[np.ndarray], np.ndarray], str],
+    other: Union[hist.BaseHist, Callable[[np.ndarray], np.ndarray]],
     likelihood: bool = False,
     *,
     ax_dict: "Optional[Dict[str, matplotlib.axes.Axes]]" = None,
+    view: str,
```

Comment:
```suggestion
    view: Literal["ratio", "view"],
```

but, better yet, can't pull just call ratio, and then modify a bit if needed, rather than having them merged like this? Having factored out functions is better, and it seems like you were mostly there.

Reply:
You'd have a plot_fit, a plot_ratio, then plot_pull would make two axes (or be given two axes) and then plot_fit on one, and plot_ratio on the other.

Reply:
Oh, I see the CI is failing as `mypy` is complaining as [`typing.Literal` was added in Python 3.8](https://docs.python.org/3/library/typing.html#typing.Literal). I assume there's some way to get things to work with older Pythons, so I'll fix this.

As for having `plot_pull` call `plot_ratio` I'm a bit confused as how that would work given that `plot_pull` and `plot_ratio` are plotting different quantities in their subplots. I guess I could see something like `plot_pull` and `plot_ratio` both calling things like `plot_fit` (and some other function to just plot the two histograms for a ratio of two hists) to populate the main axis, and then both calling a revised `plot_ratiolike` that populates the subplot with the relevant type of plot (pull or ratio).